File: haskell/diamond/src/Diamond.hs
module Diamond (diamond) where

import Data.List.Index (indexed)

makeLine :: Int -> (Int, Char) -> String
makeLine letterCount (row, letter) = 
    if letter == 'A' then outerSpaces ++ letter:outerSpaces
    else outerSpaces ++ letter:innerSpaces ++ letter:outerSpaces
    where
        outerSpaces = replicate (letterCount - row - 1) ' '
        innerSpaces = replicate (if row == 0 then 0 else row * 2 - 1) ' '

diamond :: Char -> Maybe [String]
diamond letter = Just diamondLetters
    where
        letters = indexed ['A' .. letter]
        allLetters = letters ++ (tail $ reverse letters)
        diamondLetters = map (makeLine $ length letters) allLetters

File: scala/diamond/src/main/scala/Diamond.scala
object Diamond {
  def rows(letter: Char): List[String] = {
      def makeLine(letterCount: Int, indexedLetter: (Char, Int))=  {
        val (letter, row) = indexedLetter
        val outerSpaces  = "".padTo(letterCount - row - 1, " ").mkString
        val innerSpaces = "".padTo(if (row == 0) 0 else row * 2 - 1, " ").mkString

        if (letter == 'A') s"$outerSpaces$letter$outerSpaces"
        else s"$outerSpaces$letter$innerSpaces$letter$outerSpaces"
      }

    val letters = ('A' to letter).zipWithIndex
    val allLetters = letters ++ letters.reverse.tail
    allLetters.map(makeLine(letters.length, _)).toList
  }
}
